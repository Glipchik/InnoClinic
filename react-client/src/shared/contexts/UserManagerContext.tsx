import { createContext } from "react";
import { UserManager } from 'oidc-client';

export const UserManagerContext = createContext<UserManager | null>(null);