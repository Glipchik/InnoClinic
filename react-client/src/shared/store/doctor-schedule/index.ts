import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AxiosResponse, AxiosError } from 'axios';
import { takeLatest } from 'redux-saga/effects';
import { call, CallEffect, put, PutEffect } from 'redux-saga/effects';
import { GET } from '../../api/doctor-schedule';
import TimeSlot from '../../../entities/timeSlot';
import { AnyAction } from 'redux-saga';

interface DoctorScheduleState {
  loading: boolean
  error?: string
  data?: TimeSlot[] 
}

const initialState : DoctorScheduleState = {
  loading: false,
  error: undefined,
  data: undefined
};

const fetchDoctorScheduleSlice = createSlice({
  name: 'FetchDoctorScheduleSlice',
  initialState,
  reducers: {
    fetchDoctorScheduleRequest: (state, action: PayloadAction<{ doctorId: string, date: Date }>) => {
      state.loading = true;
    },
    fetchDoctorScheduleSuccess: (state, action: PayloadAction<TimeSlot[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchDoctorScheduleFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

type ApiResponse = AxiosResponse<TimeSlot[]>;

type ApiError = AxiosError<{
  detail?: string;
  title?: string;
}>;

function* fetchDoctorSchedule(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const { doctorId, date } = action.payload;
    const response = yield call(GET, doctorId, date);
    yield put(fetchDoctorScheduleSlice.actions.fetchDoctorScheduleSuccess(response.data))
  } catch (error) {
    let errorMessage = "An unknown error occurred";

    if ((error as ApiError).response && (error as ApiError).response?.data) {
      const problemDetails = (error as ApiError).response!.data;
      errorMessage = problemDetails.detail || problemDetails.title || errorMessage;
    }
    yield put(fetchDoctorScheduleSlice.actions.fetchDoctorScheduleFailure(errorMessage))
  }
}

function* watchFetchDoctorSchedule() {
  yield takeLatest(fetchDoctorScheduleSlice.actions.fetchDoctorScheduleRequest.type, fetchDoctorSchedule);
}

export const { fetchDoctorScheduleRequest } = fetchDoctorScheduleSlice.actions;
export const fetchDoctorScheduleSliceReducer = fetchDoctorScheduleSlice.reducer
export { watchFetchDoctorSchedule }