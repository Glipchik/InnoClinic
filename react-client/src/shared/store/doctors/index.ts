import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AxiosResponse, AxiosError } from 'axios';
import { takeLatest } from 'redux-saga/effects';
import { call, CallEffect, put, PutEffect } from 'redux-saga/effects';
import doctorsApi from '../../api/doctors';
import { AnyAction } from 'redux-saga';
import DoctorModel from '../../api/doctors/models/doctorModel';

interface DoctorsState {
  loading: boolean
  error?: string
  data?: DoctorModel[]
}

const initialState : DoctorsState = {
  loading: false,
  error: undefined,
  data: undefined
};

const fetchDoctorsSlice = createSlice({
  name: 'FetchDoctorsSlice',
  initialState,
  reducers: {
    fetchDoctorsRequest: (state, action: PayloadAction<{ specializationId: string }>) => {
      state.loading = !!action;
    },
    fetchDoctorsSuccess: (state, action: PayloadAction<DoctorModel[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchDoctorsFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

type ApiResponse = AxiosResponse<DoctorModel[]>;

type ApiError = AxiosError<{
  detail?: string;
  title?: string;
}>;

function* fetchDoctors(action: AnyAction) : Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const {specializationId} = action.payload
    const response : ApiResponse  = yield call(doctorsApi.GETAll, specializationId);
    yield put(fetchDoctorsSlice.actions.fetchDoctorsSuccess(response.data))
  } catch (error) {
    let errorMessage = "An unknown error occurred";

    if ((error as ApiError).response && (error as ApiError).response?.data) {
      const problemDetails = (error as ApiError).response!.data;
      errorMessage = problemDetails.detail || problemDetails.title || errorMessage;
    }
    yield put(fetchDoctorsSlice.actions.fetchDoctorsFailure(errorMessage))
  }
}

function* watchFetchDoctors() {
  yield takeLatest(fetchDoctorsSlice.actions.fetchDoctorsRequest.type, fetchDoctors);
}

export const { fetchDoctorsRequest } = fetchDoctorsSlice.actions;
export const fetchDoctorsSliceReducer = fetchDoctorsSlice.reducer
export { watchFetchDoctors }