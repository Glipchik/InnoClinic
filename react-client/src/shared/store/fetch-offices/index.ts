import officesApi from '@shared/api/offices';
import OfficeModel from '@shared/models/offices/officeModel';
import PaginatedList from '@models/paginatedList';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface FetchOfficesState {
  loading: boolean
  error?: string
  data?: PaginatedList<OfficeModel>
}

const initialState : FetchOfficesState = {
  loading: false
};

const fetchOfficesSlice = createSlice({
  name: 'FetchOfficesSlice',
  initialState,
  reducers: {
    fetchOfficesRequest: (state, action: PayloadAction<{ pageIndex?: number, pageSize?: number }>) => {
      state.loading = !!action;
    },
    fetchOfficesSuccess: (state, action: PayloadAction<PaginatedList<OfficeModel>>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchOfficesFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchOfficesRequest } = fetchOfficesSlice.actions;
export const fetchOfficesSliceReducer = fetchOfficesSlice.reducer

type ApiResponse = AxiosResponse<PaginatedList<OfficeModel>>;

function* fetchOffices(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const { pageIndex, pageSize } = action.payload;
    const response: ApiResponse = yield call(officesApi.getAll, pageIndex, pageSize);
    yield put(fetchOfficesSlice.actions.fetchOfficesSuccess(response.data))
  } catch (error) {
    yield put(fetchOfficesSlice.actions.fetchOfficesFailure(handleError(error as ApiError)))
  }
}

function* watchFetchOffices() {
  yield takeLatest(fetchOfficesRequest.type, fetchOffices);
}

export { watchFetchOffices }