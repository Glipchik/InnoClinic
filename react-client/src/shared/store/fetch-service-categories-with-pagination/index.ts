import PaginatedList from '@models/paginatedList';
import ServiceCategoryModel from '@models/serviceCategories/serviceCategoryModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import serviceCategoriesApi from '@shared/api/service-categories';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface FetchServiceCategoriesWithPaginationState {
  loading: boolean
  error?: string
  data?: PaginatedList<ServiceCategoryModel>
}

const initialState : FetchServiceCategoriesWithPaginationState = {
  loading: false
};

const fetchServiceCategoriesWithPaginationSlice = createSlice({
  name: 'FetchServiceCategoriesWithPaginationSlice',
  initialState,
  reducers: {
    fetchServiceCategoriesWithPaginationRequest: (state, action: PayloadAction<{ pageIndex?: number, pageSize?: number }>) => {
      state.loading = !!action;
    },
    fetchServiceCategoriesWithPaginationSuccess: (state, action: PayloadAction<PaginatedList<ServiceCategoryModel>>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchServiceCategoriesWithPaginationFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchServiceCategoriesWithPaginationRequest } = fetchServiceCategoriesWithPaginationSlice.actions;
export const fetchServiceCategoriesWithPaginationSliceReducer = fetchServiceCategoriesWithPaginationSlice.reducer

type ApiResponse = AxiosResponse<PaginatedList<ServiceCategoryModel>>;

function* fetchServiceCategoriesWithPagination(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const { pageIndex, pageSize } = action.payload;
    const response: ApiResponse = yield call(serviceCategoriesApi.getAllWithPagination, pageIndex, pageSize);
    yield put(fetchServiceCategoriesWithPaginationSlice.actions.fetchServiceCategoriesWithPaginationSuccess(response.data))
  } catch (error) {
    yield put(fetchServiceCategoriesWithPaginationSlice.actions.fetchServiceCategoriesWithPaginationFailure(handleError(error as ApiError)))
  }
}

function* watchFetchServiceCategoriesWithPagination() {
  yield takeLatest(fetchServiceCategoriesWithPaginationRequest.type, fetchServiceCategoriesWithPagination);
}

export { watchFetchServiceCategoriesWithPagination }