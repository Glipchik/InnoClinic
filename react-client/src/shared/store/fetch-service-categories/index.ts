import ServiceCategoryModel from '@models/serviceCategories/serviceCategoryModel';
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import serviceCategoriesApi from '@shared/api/service-categories';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface FetchServiceCategoriesState {
  loading: boolean
  error?: string
  data?: ServiceCategoryModel[]
}

const initialState : FetchServiceCategoriesState = {
  loading: false
};

const fetchServiceCategoriesSlice = createSlice({
  name: 'FetchServiceCategoriesSlice',
  initialState,
  reducers: {
    fetchServiceCategoriesRequest: (state) => {
      state.loading = true;
    },
    fetchServiceCategoriesSuccess: (state, action: PayloadAction<ServiceCategoryModel[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchServiceCategoriesFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchServiceCategoriesRequest } = fetchServiceCategoriesSlice.actions;
export const fetchServiceCategoriesSliceReducer = fetchServiceCategoriesSlice.reducer

type ApiResponse = AxiosResponse<ServiceCategoryModel[]>;

function* fetchServiceCategories(): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const response: ApiResponse = yield call(serviceCategoriesApi.getAll);
    yield put(fetchServiceCategoriesSlice.actions.fetchServiceCategoriesSuccess(response.data))
  } catch (error) {
    yield put(fetchServiceCategoriesSlice.actions.fetchServiceCategoriesFailure(handleError(error as ApiError)))
  }
}

function* watchFetchServiceCategories() {
  yield takeLatest(fetchServiceCategoriesRequest.type, fetchServiceCategories);
}

export { watchFetchServiceCategories }