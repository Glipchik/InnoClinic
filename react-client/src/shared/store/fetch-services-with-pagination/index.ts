import servicesApi from '@shared/api/services';
import PaginatedList from '@models/paginatedList';
import ServiceModel from '@models/services/serviceModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface FetchServicesWithPaginationState {
  loading: boolean
  error?: string
  data?: PaginatedList<ServiceModel>
}

const initialState : FetchServicesWithPaginationState = {
  loading: false
};

const fetchServicesWithPaginationSlice = createSlice({
  name: 'FetchServicesWithPaginationSlice',
  initialState,
  reducers: {
    fetchServicesWithPaginationRequest: (state, action: PayloadAction<{ pageIndex?: number, pageSize?: number }>) => {
      state.loading = !!action;
    },
    fetchServicesWithPaginationSuccess: (state, action: PayloadAction<PaginatedList<ServiceModel>>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchServicesWithPaginationFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchServicesWithPaginationRequest } = fetchServicesWithPaginationSlice.actions;
export const fetchServicesWithPaginationSliceReducer = fetchServicesWithPaginationSlice.reducer

type ApiResponse = AxiosResponse<PaginatedList<ServiceModel>>;

function* fetchServicesWithPagination(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const { pageIndex, pageSize } = action.payload;
    const response: ApiResponse = yield call(servicesApi.getAllWithPagination, pageIndex, pageSize);
    yield put(fetchServicesWithPaginationSlice.actions.fetchServicesWithPaginationSuccess(response.data))
  } catch (error) {
    yield put(fetchServicesWithPaginationSlice.actions.fetchServicesWithPaginationFailure(handleError(error as ApiError)))
  }
}

function* watchFetchServicesWithPagination() {
  yield takeLatest(fetchServicesWithPaginationRequest.type, fetchServicesWithPagination);
}

export { watchFetchServicesWithPagination }