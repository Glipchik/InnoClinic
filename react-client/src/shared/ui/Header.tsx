import { useContext, useEffect, useState } from "react";
import { UserManagerContext } from '../contexts/UserManagerContext';
import { User } from 'oidc-client';
import { Link } from "react-router-dom";
import { useSelector } from 'react-redux';
import Button from "./controls/Button";
import React from "react";
import { RootState } from "../../store/store";

function Header() {
  const userManager = useContext(UserManagerContext);
  const [user, setUser] = useState<User | null>(null);

  const { isUserAuthorized } = useSelector(
    (state: RootState) => state.auth
  );

  useEffect(() => {
    if (userManager) {
      async function fetchUser() {
        const user = await userManager!.getUser();
        setUser(user);
      }
      fetchUser();
    }
  }, [userManager, isUserAuthorized]);

  return (
    <header className="w-full inline-flex py-7 px-3 items-center bg-gray-700 text-white">
      <div className="w-[25%] h-full">
        <h1 className="text-5xl font-bold">InnoClinic</h1>
      </div>
      <div className="w-[45%] h-full">

      </div>
      <div className="w-[30%] h-full">
        <div className="flex w-full h-full items items-center space-x-3 justify-end">
          {user ? (
            <>
              <div className="flex-col h-full space-y-0">
                <p className="text-xl">
                  {user.profile.email}
                </p>
                <p className="text-sm text-gray-300">
                  {user.profile.role}
                </p>
              </div>
              <li className="list-none"> <Link to="/logout"> <Button className="bg-red-800 hover:bg-red-900"> Logout </Button> </Link> </li>
            </>
          ) : (
            <>
              <li className="list-none"> <Link to="/login"> <Button> Login </Button> </Link> </li>
              <li className="list-none">  <Link to="/register"> <Button> Registration </Button> </Link> </li>
            </>
          )}
        </div>
      </div>
    </header>
  );
}

export default Header;