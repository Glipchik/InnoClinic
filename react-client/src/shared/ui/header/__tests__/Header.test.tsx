import { MemoryRouter, Route, Routes } from "react-router-dom";
import Header from "@shared/ui/header";
import store from "@app/store";
import { UserManagerContext } from "@shared/contexts/UserManagerContext";
import { Provider } from "react-redux";
import { UserManager, WebStorageStateStore } from "oidc-client";
import { mockUserManager } from "./__mocks__";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

const config = {
  authority: "",
  client_id: "",
  redirect_uri: "",
  response_type: "",
  scope: "",
  post_logout_redirect_uri: "",
  userStore: new WebStorageStateStore({ store: window.localStorage }),
  stateStore: new WebStorageStateStore({ store: window.sessionStorage }),
  loadUserInfo: true,
  automaticSilentRenew: true,
  silent_redirect_uri: "",
  monitorSession: false,
};

const userManager = new UserManager(config);

const renderHeader = (userManager: UserManager) =>
  render(
    <Provider store={store}>
      <UserManagerContext.Provider value={userManager}>
        <MemoryRouter>
          <Routes>
            <Route
              path="/login"
              element={<h1 data-testid="login-page">Login</h1>}
            />
            <Route
              path="/logout"
              element={<h1 data-testid="logout-page">Logout</h1>}
            />
            <Route path="/" element={<Header />} />
          </Routes>
        </MemoryRouter>
      </UserManagerContext.Provider>
    </Provider>
  );

describe("Header", () => {
  it("renders required components when unauthorized", async () => {
    const { getByTestId, queryByTestId } = renderHeader(userManager);

    await waitFor(() => {
      expect(queryByTestId("email")).not.toBeInTheDocument();
      expect(queryByTestId("logout-button")).not.toBeInTheDocument();
      expect(queryByTestId("logo")).toBeInTheDocument();
    });

    const signInUpButton = getByTestId("sign-in-up-button");
    expect(signInUpButton).toBeInTheDocument();
    await userEvent.click(signInUpButton);

    await waitFor(() => {
      expect(queryByTestId("login-page")).toBeInTheDocument();
    });
  });

  it("renders required components when authorized", async () => {
    const { getByTestId, queryByTestId, queryByText } =
      renderHeader(mockUserManager);

    await waitFor(() => {
      const emailComponent = queryByTestId("email");
      expect(emailComponent).toBeInTheDocument();
      expect(emailComponent).toHaveTextContent("test@example.com");
      expect(queryByText("Receptionist")).toBeInTheDocument();
      expect(queryByTestId("logout-button")).toBeInTheDocument();
      expect(queryByTestId("logo")).toBeInTheDocument();
    });

    const logoutButton = getByTestId("logout-button");
    expect(logoutButton).toBeInTheDocument();
    await userEvent.click(logoutButton);

    await waitFor(() => {
      expect(queryByTestId("logout-page")).toBeInTheDocument();
    });
  });

  it("matches snapshot", () => {
    const { asFragment } = renderHeader(mockUserManager);
    expect(asFragment()).toMatchSnapshot();
  });
});
