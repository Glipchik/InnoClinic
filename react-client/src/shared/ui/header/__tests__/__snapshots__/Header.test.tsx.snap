// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="w-full inline-flex py-7 px-3 items-center bg-gray-100 text-black"
  >
    <div
      class="w-[25%] h-full"
      data-testid="logo"
    >
      <h1
        class="text-5xl font-bold"
      >
        InnoClinic
      </h1>
    </div>
    <div
      class="w-[45%] h-full"
    />
    <div
      class="w-[30%] h-full"
    >
      <div
        class="flex w-full h-full items items-center space-x-3 justify-end"
      >
        <li
          class="list-none"
        >
          <a
            data-discover="true"
            href="/login"
          >
            <button
              class="px-4 py-2 bg-blue-500 text-white rounded disabled:bg-blue-900 disabled:text-gray-400 disabled:hover:cursor-not-allowed hover:bg-blue-600 focus:outline-none cursor-pointer undefined"
              data-testid="sign-in-up-button"
            >
               Sign in/up 
            </button>
          </a>
        </li>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
